src/main/java/de/tum/in/www1/artemis/service/connectors/BitbucketUserManagementService.java

```diff
@@ -0,0 +1,179 @@
+     * Deactivates the VCS user.
+     *
+     * @param login The username of the user in the VCS
+     * @throws VersionControlException if an exception occurred
+     */
+    @Override
+    public void deactivateUser(String login) throws VersionControlException {
+        // Not supported by Bitbucket
+    }
+
+    /**
+     * Updates all exercises in a course based on the new instructors, editors and teaching assistant groups. This entails removing
+     * all users from exercises, that are no longer part of any relevant group and adding all users to exercises in the course
+     * that are part of the updated groups.
+     *
+     * @param updatedCourse             The updated course with the new permissions
+     * @param oldInstructorGroup        The old instructor group name
+     * @param oldEditorGroup            The old editor group name
+     * @param oldTeachingAssistantGroup The old teaching assistant group name
+     */
+    @Override
+    public void updateCoursePermissions(Course updatedCourse, String oldInstructorGroup, String oldEditorGroup, String oldTeachingAssistantGroup) {
+        if (oldInstructorGroup.equals(updatedCourse.getInstructorGroupName()) && oldEditorGroup.equals(updatedCourse.getEditorGroupName())
+                && oldTeachingAssistantGroup.equals(updatedCourse.getTeachingAssistantGroupName())) {
+            // Do nothing if the group names didn't change
+            return;
+        }
+
+        final List<ProgrammingExercise> programmingExercises = programmingExerciseRepository.findAllProgrammingExercisesInCourseOrInExamsOfCourse(updatedCourse);
+        log.info("Update Bitbucket permissions for programming exercises: " + programmingExercises.stream().map(ProgrammingExercise::getProjectKey).toList());
+
+        for (ProgrammingExercise programmingExercise : programmingExercises) {
+            if (!oldInstructorGroup.equals(updatedCourse.getInstructorGroupName())) {
+                bitbucketService.grantGroupPermissionToProject(programmingExercise.getProjectKey(), oldInstructorGroup, null);
+                bitbucketService.grantGroupPermissionToProject(programmingExercise.getProjectKey(), updatedCourse.getInstructorGroupName(), BitbucketPermission.PROJECT_ADMIN);
+            }
+            if (!oldEditorGroup.equals(updatedCourse.getEditorGroupName())) {
+                bitbucketService.grantGroupPermissionToProject(programmingExercise.getProjectKey(), oldEditorGroup, null);
+                bitbucketService.grantGroupPermissionToProject(programmingExercise.getProjectKey(), updatedCourse.getEditorGroupName(), BitbucketPermission.PROJECT_WRITE);
+            }
+            boolean isCourseOrAfterExam = programmingExercise.isCourseExercise() || ZonedDateTime.now().isAfter(programmingExercise.getExerciseGroup().getExam().getStartDate()
+                    .plusMinutes(programmingExercise.getExerciseGroup().getExam().getStudentExams().stream().mapToInt(StudentExam::getWorkingTime).max().orElse(0)));
```

Comment:
The calculation for `boolean isAfterLatestStudentExamEnd(programmingExercise.getExerciseGroup().getExam())` could maybe be useful in other places as well. I think it could be added to the Exam class with a `@JsonIgnore` annotation.

Reply:
Done, please recheck

Reply:
FIne for me.